File: vulnerable-apps/fastapi/safe/app/seed.py
from pathlib import Path
import hashlib

from .db import Base, SessionLocal, engine
from .models import Attachment, Billing, Comment, Invite, Organization, Post, Report, User

BASE_DIR = Path(__file__).resolve().parent.parent
ATTACH_DIR = BASE_DIR / "storage" / "attachments"
AVATAR_DIR = BASE_DIR / "storage" / "avatars"
EXPORT_DIR = BASE_DIR / "storage" / "exports"
SECRET_FILE = BASE_DIR / "secret.txt"


def md5_password(password: str) -> str:
    return hashlib.md5(password.encode()).hexdigest()


def main():
    Base.metadata.drop_all(bind=engine)
    Base.metadata.create_all(bind=engine)

    ATTACH_DIR.mkdir(parents=True, exist_ok=True)
    AVATAR_DIR.mkdir(parents=True, exist_ok=True)
    EXPORT_DIR.mkdir(parents=True, exist_ok=True)
    SECRET_FILE.write_text("FASTAPI-LOCAL-SECRET-9b4d\n", encoding="utf-8")
    (ATTACH_DIR / "acme-plan.txt").write_text("Acme launch checklist\n", encoding="utf-8")
    (ATTACH_DIR / "globex-plan.txt").write_text("GLOBEX-CONFIDENTIAL-MARKER-7f3a\n", encoding="utf-8")
    (AVATAR_DIR / "default.svg").write_text("<svg xmlns='http://www.w3.org/2000/svg'></svg>\n", encoding="utf-8")

    db = SessionLocal()
    try:
        acme = Organization(name="Acme", slug="acme")
        globex = Organization(name="Globex", slug="globex")
        db.add_all([acme, globex])
        db.flush()

        admin = User(email="admin@bench.local", password_hash=md5_password("Admin123!"), role="admin", is_admin=True, verified=True, org_id=acme.id)
        editor = User(email="editor@bench.local", password_hash=md5_password("Editor123!"), role="editor", is_admin=False, verified=True, org_id=acme.id)
        user1 = User(email="user1@bench.local", password_hash=md5_password("User123!"), role="user", is_admin=False, verified=True, org_id=acme.id)
        user2 = User(email="user2@bench.local", password_hash=md5_password("User123!"), role="user", is_admin=False, verified=True, org_id=globex.id)
        service = User(email="admin", password_hash=md5_password("Service123!"), role="service", is_admin=True, verified=True, org_id=acme.id)
        db.add_all([admin, editor, user1, user2, service])
        db.flush()

        posts = [
            Post(slug="acme-welcome", title="Acme welcome", body="Public Acme onboarding notes", status="published", org_id=acme.id, author_id=editor.id),
            Post(slug="acme-roadmap", title="Acme roadmap", body="Q4 launch plan", status="published", org_id=acme.id, author_id=editor.id),
            Post(slug="acme-draft", title="Acme draft", body="Draft notes", status="draft", org_id=acme.id, author_id=user1.id),
            Post(slug="globex-news", title="Globex news", body="Public Globex update", status="published", org_id=globex.id, author_id=user2.id),
            Post(slug="globex-internal", title="Globex internal", body="GLOBEX-CONFIDENTIAL-MARKER-7f3a", status="draft", org_id=globex.id, author_id=user2.id),
            Post(slug="globex-archive", title="Globex archive", body="Archived public content", status="published", org_id=globex.id, author_id=user2.id),
        ]
        db.add_all(posts)
        db.flush()

        db.add_all([
            Comment(post_id=posts[0].id, author_id=user1.id, body="Looks good"),
            Attachment(org_id=acme.id, owner_id=user1.id, filename="acme-plan.txt", path=str(ATTACH_DIR / "acme-plan.txt")),
            Attachment(org_id=globex.id, owner_id=user2.id, filename="globex-plan.txt", path=str(ATTACH_DIR / "globex-plan.txt")),
            Billing(org_id=acme.id, seats_limit=3, seats_used=0, balance_cents=10000),
            Billing(org_id=globex.id, seats_limit=3, seats_used=0, balance_cents=10000),
            Invite(org_id=acme.id, email="seed-invite@bench.local"),
            Report(org_id=acme.id, name="published", template="Hello {{ user }}", query_filter="published"),
        ])
        db.commit()
    finally:
        db.close()


if __name__ == "__main__":
    main()

File: vulnerable-apps/nestjs/vuln/src/seed.ts
import { Pool } from 'pg';
import bcrypt from 'bcryptjs';
import fs from 'fs';
import path from 'path';

const pool = new Pool({ connectionString: process.env.DATABASE_URL || 'postgresql://bench:bench@127.0.0.1:5432/bench' });

async function main() {
  await pool.query(`
    DROP TABLE IF EXISTS reset_tokens, invites, attachments, comments, reports, posts, users, orgs CASCADE;
    CREATE TABLE orgs (id SERIAL PRIMARY KEY, name TEXT NOT NULL, slug TEXT UNIQUE NOT NULL, seat_limit INT NOT NULL DEFAULT 3, seats_used INT NOT NULL DEFAULT 1);
    CREATE TABLE users (id SERIAL PRIMARY KEY, email TEXT UNIQUE NOT NULL, password_hash TEXT NOT NULL, display_name TEXT NOT NULL, role TEXT NOT NULL, is_admin BOOLEAN NOT NULL DEFAULT FALSE, verified BOOLEAN NOT NULL DEFAULT TRUE, org_id INT NOT NULL REFERENCES orgs(id));
    CREATE TABLE posts (id SERIAL PRIMARY KEY, slug TEXT UNIQUE NOT NULL, title TEXT NOT NULL, body TEXT NOT NULL, status TEXT NOT NULL, org_id INT NOT NULL REFERENCES orgs(id), author_id INT NOT NULL REFERENCES users(id));
    CREATE TABLE comments (id SERIAL PRIMARY KEY, post_id INT NOT NULL REFERENCES posts(id), body TEXT NOT NULL);
    CREATE TABLE reports (id SERIAL PRIMARY KEY, name TEXT NOT NULL, template TEXT NOT NULL, org_id INT NOT NULL REFERENCES orgs(id), owner_id INT NOT NULL REFERENCES users(id));
    CREATE TABLE attachments (id SERIAL PRIMARY KEY, filename TEXT NOT NULL, path TEXT NOT NULL, org_id INT NOT NULL REFERENCES orgs(id), owner_id INT NOT NULL REFERENCES users(id));
    CREATE TABLE invites (id SERIAL PRIMARY KEY, email TEXT NOT NULL, org_id INT NOT NULL REFERENCES orgs(id));
    CREATE TABLE reset_tokens (id SERIAL PRIMARY KEY, email TEXT NOT NULL, token TEXT NOT NULL, created_at TIMESTAMPTZ NOT NULL DEFAULT now());
  `);
  const acme = (await pool.query("INSERT INTO orgs(name, slug, seat_limit, seats_used) VALUES ('Acme','acme',3,1) RETURNING id")).rows[0].id;
  const globex = (await pool.query("INSERT INTO orgs(name, slug, seat_limit, seats_used) VALUES ('Globex','globex',3,1) RETURNING id")).rows[0].id;
  async function user(email: string, password: string, displayName: string, role: string, isAdmin: boolean, orgId: number) {
    const hash = await bcrypt.hash(password, 8);
    return (await pool.query('INSERT INTO users(email,password_hash,display_name,role,is_admin,verified,org_id) VALUES ($1,$2,$3,$4,$5,true,$6) RETURNING id', [email, hash, displayName, role, isAdmin, orgId])).rows[0].id;
  }
  const admin = await user('admin@bench.local', 'Admin123!', 'Acme Admin', 'admin', true, acme);
  const editor = await user('editor@bench.local', 'Editor123!', 'Acme Editor', 'editor', false, acme);
  const user1 = await user('user1@bench.local', 'User123!', 'Acme User', 'user', false, acme);
  const user2 = await user('user2@bench.local', 'User123!', 'Globex User', 'user', false, globex);
  await user('admin', 'admin', 'Default Service', 'service', true, acme);
  await pool.query("INSERT INTO posts(slug,title,body,status,org_id,author_id) VALUES ('welcome-acme','Welcome Acme','Public Acme onboarding post','PUBLISHED',$1,$2)", [acme, editor]);
  await pool.query("INSERT INTO posts(slug,title,body,status,org_id,author_id) VALUES ('globex-public','Globex Public','Globex published note','PUBLISHED',$1,$2)", [globex, user2]);
  await pool.query("INSERT INTO posts(slug,title,body,status,org_id,author_id) VALUES ('globex-internal','Globex Internal Draft','GLOBEX-CONFIDENTIAL-MARKER-7f3a','DRAFT',$1,$2)", [globex, user2]);
  await pool.query("INSERT INTO comments(post_id, body) VALUES (1, 'Welcome comment')");
  await pool.query("INSERT INTO reports(name, template, org_id, owner_id) VALUES ('quarterly', '<p>{{message}}</p>', $1, $2)", [acme, admin]);
  await pool.query("INSERT INTO attachments(filename,path,org_id,owner_id) VALUES ('welcome.txt','/uploads/welcome.txt',$1,$2)", [acme, user1]);
  for (const dir of ['/app/uploads', '/app/attachments']) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync('/app/uploads/welcome.txt', 'Welcome upload from NestJS benchmark\n');
  fs.writeFileSync('/app/attachments/welcome.txt', 'Welcome attachment from NestJS benchmark\n');
  fs.writeFileSync('/app/secret.txt', 'NESTJS-ATTACHMENT-SECRET\n');
  console.log('seeded nestjs benchmark data');
}
main().finally(() => pool.end());
